extensions/message-redaction.md

```diff
@@ -0,0 +1,191 @@
+---
+title: Message redaction
+layout: spec
+work-in-progress: true
+copyrights:
+  -
+    name: "James Wheare"
+    email: "james@irccloud.com"
+    period: "2020"
+  -
+    name: "Val Lorentz"
+    email: "progval+ircv3@progval.net"
+    period: "2023"
+---
+
+## Notes for implementing work-in-progress version
+
+This is a work-in-progress specification.
+
+Software implementing this work-in-progress specification MUST NOT use the
+unprefixed `message-redaction` capability name.
+Instead, implementations SHOULD use the `draft/message-redaction` capability
+name to be interoperable with other software implementing a compatible
+work-in-progress version.
+
+The final version of the specification will use an unprefixed capability name.
+
+## Introduction
+
+This specification enables messages to be deleted.
+Use cases include retracting accidentally sent messages, moderation,
+and removing a [`+draft/react` client tag][], amongst others.
+These are cosmetic use cases and do not provide any operational security
+guarantees.
+
+## Architecture
+
+### Dependencies
+
+Clients wishing to use these capabilities MUST negotiate the [`message-tags`][]
+capability with the server.
+Additionally, this capability relies on messages being sent with the
+[`msgid`][] tag, and message ids MUST NOT contain spaces.
+Clients SHOULD negotiate the [`echo-message`][] capabilities in order to receive
+message IDs for their own messages, to allow them to be redacted.
+
+
+### Capability
+
+This specification adds the `draft/message-redaction` capability.
+Clients MUST ignore this capability's value, if any.
+
+Implementations that negotiate these capabilities indicate that they are
+capable of handling the respective commands and message tags described below.
+
+
+### Command
+
+To redact a message, a client MUST negotiate the `draft/message-redaction`
+capability and send a `REDACT` command to a target nickname or channel.
+The command is defined as follows:
+
+    REDACT <target> <msgid> [:reason]
+
+Where `<msgid>` is the id of the message to be redacted.
+
+If the client is authorised to delete the message, the server:
+
+* SHOULD forward this `REDACT`, with an appropriate prefix, to the target
+  recipients which negotiated the `draft/message-redaction` capability, in the
+  same way as PRIVMSG messages.
+* MUST not forward this `REDACT` to target recipients which did not negotiated
+  this capability (see "Fallback" below)
+
+### Chat history
+
+After a message is redacted, [`chathistory`][] responses SHOULD either:
+
+* exclude it entirely
+* replace its second parameter and/or tags with a placeholder and
+  add the `REDACT` message to the response (not counting toward message limits)
+* add the `REDACT` message to the response (not counting toward message limits)
```

Comment:
Why do client need to differentiate between the two?

Reply:
Per the linked discussion above, my clients check whether there is more history to fetch by counting the number of messages. If it's less than the limit, no more messages.

Hm, but now that I've written this, this would only cause the number of messages to exceed the limit, causing needless but harmless chathistory requests in the edge case where there are no more messages. So it should be all fine?

Does a standalone `REDACT` count towards the limit? I'd say yes. This draws a line between "standalone" and "context" messages.

Reply:
The discussion hasn't really settled AFAIU, and my clients are still doing the limit check.

Please note, there is an interleaved discussion about what should happen when a channel has chathistory disabled, and this one has settled on empty batches instead of a `FAIL`. But it's unrelated.

Reply:
Oh, and another reason why clients must be able to tell the difference between a "context" `REDACT` and a "standalone" one: timestamp-based chathistory queries. Let's say a batch ends with a deleted message:

```
@time=2022-01-01T20:00:00Z;msgid=foo PRIVMSG #chan :Hey there
@time=2023-05-01T20:00:00Z REDACT #chan foo
```

Then the next chathistory request will fetch all messages after 2023-05-01…

Reply:
There are two ways to fix this issue that I see:

- Add a `@chathistory-context` tag to all "related"/"context" messages returned by chathistory. Clients can ignore these when figuring out the selector for their next query.
- Add cursor-based paging to chathistory: #524.

Reply:
Does a resolution to this issue require changes to the redaction spec, or just to the chathistory spec? i.e. is this a blocker for merge/ratification?

Reply:
only the chathistory spec IMO

Reply:
Just spotted this other comment from @emersion that appears unaddressed https://github.com/ircv3/ircv3-specifications/pull/524/files#r1213547049

This PR has become a bit of a nightmare to review lol.